core/trino-main/src/main/java/io/trino/execution/SqlTaskManager.java

```diff
@@ -597,4 +650,87 @@ public QueryContext getQueryContext(QueryId queryId)
+    }
+
+    @VisibleForTesting
+    public void failStuckSplitTasks()
+    {
+        stuckSplitTaskInterrupter.ifPresent(StuckSplitTaskInterrupter::failStuckSplitTasks);
+    }
+
+    private Optional<StuckSplitTaskInterrupter> createStuckSplitTaskInterrupter(
+            boolean enableInterruptStuckSplitTasks,
+            Duration stuckSplitsWarningThreshold,
+            Duration interruptStuckSplitTasksTimeout,
+            Duration stuckSplitsDetectionInterval,
+            Predicate<List<StackTraceElement>> stuckSplitStackTracePredicate,
+            TaskExecutor taskExecutor,
+            Ticker ticker)
+    {
+        if (!enableInterruptStuckSplitTasks) {
+            return Optional.empty();
+        }
+        return Optional.of(new StuckSplitTaskInterrupter(stuckSplitsWarningThreshold, interruptStuckSplitTasksTimeout, stuckSplitsDetectionInterval, stuckSplitStackTracePredicate, taskExecutor, ticker));
+    }
+
+    private class StuckSplitTaskInterrupter
+    {
+        private final Duration interruptStuckSplitTasksTimeout;
+        private final Duration stuckSplitsDetectionInterval;
+        private Predicate<List<StackTraceElement>> stuckSplitStackTracePredicate;
+        private final TaskExecutor taskExecutor;
+        private final Ticker ticker;
+
+        public StuckSplitTaskInterrupter(
+                Duration stuckSplitsWarningThreshold,
+                Duration interruptStuckSplitTasksTimeout,
+                Duration stuckSplitDetectionInterval,
+                Predicate<List<StackTraceElement>> stuckSplitStackTracePredicate,
+                TaskExecutor taskExecutor,
+                Ticker ticker)
+        {
+            checkArgument(interruptStuckSplitTasksTimeout.compareTo(SPLIT_RUN_QUANTA) >= 0, "interruptStuckSplitTasksTimeout must be at least %s", SPLIT_RUN_QUANTA);
+            checkArgument(stuckSplitsWarningThreshold.compareTo(interruptStuckSplitTasksTimeout) <= 0, "interruptStuckSplitTasksTimeout cannot be less than stuckSplitsWarningThreshold");
+
+            this.interruptStuckSplitTasksTimeout = requireNonNull(interruptStuckSplitTasksTimeout, "interruptStuckSplitTasksTimeout is null");
+            this.stuckSplitsDetectionInterval = requireNonNull(stuckSplitDetectionInterval, "stuckSplitsDetectionInterval is null");
+            this.stuckSplitStackTracePredicate = requireNonNull(stuckSplitStackTracePredicate, "stuckSplitStackTracePredicate is null");
+            this.taskExecutor = requireNonNull(taskExecutor, "taskExecutor is null");
+            this.ticker = requireNonNull(ticker, "ticker is null");
+        }
+
+        public Duration getStuckSplitsDetectionInterval()
+        {
+            return stuckSplitsDetectionInterval;
+        }
+
+        private void failStuckSplitTasks()
+        {
+            long currentTime = ticker.read();
+            Set<TaskId> stuckSplitTaskIds = this.taskExecutor.getStuckSplitTaskIds((RunningSplitInfo splitInfo) -> {
```

Comment:
nit: `this.` is not needed (we prefer not to specify `this` if it is not needed)